### ISSUE TITLE
不相关的句子的相似度得分很高

### ISSUE THREAD

**Issue by goggeryang** [FlagOpen/FlagEmbedding#1]
几个毫不相关的句子的得分基本上都有0.7,0.8左右，那相关性高的阈值是多少，为什么会存在不相干的相似度很高的情况

**Comment by staoxiao** [comment]
大部分bert系向量模型的基本相似度都会比较高，您需要根据实际情况设置阈值，0.7-0.8并不能代表相关。

**Comment by staoxiao** [comment]
方便的话，您可以提供一下使用的句子和模型，以进一步判断。

**Comment by goggeryang** [comment]
模型：BAAI/bge-large-zh
句或词
string1[6]: 证券事务代表
string2[2]: 券事
相似度: 0.8983931541442871

string1[6]: 证券事务代表
string2[3]: 券 事
相似度: 0.8983931541442871

string1[6]: 证券事务代表
string2[2]: 务代
相似度: 0.8694437742233276

string1[6]: 证券事务代表
string2[3]: 务 代
相似度: 0.8694437742233276

string1[6]: 证券事务代表
string2[1]:  
相似度: 0.8166958689689636

**Comment by staoxiao** [comment]
收到，感谢！
string1[6]: 证券事务代表
string2[2]: 证券代表
相似度: 0.96933395

string1[6]: 证券事务代表
string2[2]: 证券事务
相似度: 0.9447619

`证券代表`和`证券事务`的相似度明显高于其他句子，是满足了bge训练的目标的(bge训练目标并不是二分类，而是对比排序学习）。这种短文本匹配场景建议根据实际数据选择阈值。

**Comment by goggeryang** [comment]
好的，了解

**Comment by ReverseSystem001** [comment]
为什么使用FlagModel 相似度跟你们的不一样？ 下面是运行代码：
from FlagEmbedding import FlagModel
import numpy as np

sentences = ["证券事务代表"]
model = FlagModel('/data/ywb/LLM_Datas/model/EmbeddingModels/bge-large-zh', query_instruction_for_retrieval="为这个句子生成表示以用于检索相关文章：")
embeddings = model.encode(sentences)
print(embeddings)

# 对于检索任务中的查询问题，请使用 encode_queries() 函数，其会自动为每个查询加上指令
# 检索中的候选集依然使用 encode() 或 encode_corpus() 函数
queries = ['证券事务代表', "这是啥"]
passages = ["券事", "务代", "证券代表", "证券事务"]
q_embeddings = model.encode_queries(queries)
p_embeddings = model.encode(passages)
scores = q_embeddings @ p_embeddings.T
max_index = np.argmax(scores, axis=1)


print(type(scores))
print(scores)
print("每行的最大值对应的索引：", max_index)

results:
----------using 2*GPUs----------
[[ 0.01721212 -0.02396812 -0.032729   ...  0.02451125  0.00439007
  -0.02822394]]
<class 'numpy.ndarray'>
[[0.6954609  0.6524521  0.7475003  0.7268134 ]
 [0.58254194 0.592916   0.5847791  0.58445   ]]

**Comment by staoxiao** [comment]
您好，短文本匹配不建议使用encode_queries， 直接使用encode就好

**Comment by ReverseSystem001** [comment]
嗯嗯，试了一下。分数提高了，同一个模型为啥会分短文本匹配和长文本匹配？

**Comment by staoxiao** [comment]
encode_queries会为query增加指令，适用于使用短query去检索较长文档。
其他任务由于训练中都没有增加指令，测试与训练保持一致效果最好。
并且由于两个短文本本身词汇就少，加上指令对生成的向量变化较大，因此要么就两个文本都加上，要么就都不加。

**Comment by staoxiao** [comment]
用户问题使用encode_queries，本地知识使用encode

**Comment by cristianohello** [comment]
@staoxiao 
比如说，用户问题是query：卡塔尔世界杯冠军是谁？
而本地的知识库文件为TXT文件，文件里面有大量的百度百科文本，包括体育、娱乐等等大量的纯文本。我想检索出来query相关的TXT文件中的卡塔尔世界杯相关的内容。文本内容可能有几十字，或者几百字。我该用encode_queries这个吧？但是我之前没有遇到过加上“model = FlagModel('/data/ywb/LLM_Datas/model/EmbeddingModels/bge-large-zh', query_instruction_for_retrieval="为这个句子生成表示以用于检索相关文章：")”这种指令的向量模型。检索模型为什么会有这个指令呢？

**Comment by staoxiao** [comment]
卡塔尔世界杯冠军是谁？用encode_queries， TXT文本内容用encode。
由于检索任务通常query和候选文本差异较大（比如长度上），增加指令可以让模型分清是query还是文本，区别对待query和文本的建模，效果更好一些

**Comment by cristianohello** [comment]
@staoxiao 

能举个代码的例子吗？

是这样吗
from transformers import AutoTokenizer, AutoModel
import torch

queries = ['卡塔尔世界杯冠军是谁？, '中国首都是哪']

# Sentences we want sentence embeddings for
sentences = ["卡塔尔世界杯冠军是？", "北京是中国首都"]
# Load model from HuggingFace Hub
tokenizer = AutoTokenizer.from_pretrained('BAAI/bge-large-zh')
model = AutoModel.from_pretrained('BAAI/bge-large-zh')
# Tokenize sentences
encoded_input = tokenizer(sentences, padding=True, truncation=True, return_tensors='pt')
# for retrieval task, add an instruction to query
# encoded_input = tokenizer([instruction + q for q in queries], padding=True, truncation=True, return_tensors='pt')
# Compute token embeddings
with torch.no_grad():
    model_output = model(**encoded_input)
    # Perform pooling. In this case, cls pooling.
    sentence_embeddings = model_output[0][:, 0]
# normalize embeddings
sentence_embeddings = torch.nn.functional.normalize(sentence_embeddings, p=2, dim=1)
print("Sentence embeddings:", sentence_embeddings)

**Comment by staoxiao** [comment]
是的，
对于文本使用encoded_input = tokenizer(sentences, padding=True, truncation=True, return_tensors='pt')
对于查询使用encoded_input = tokenizer([instruction + q for q in queries], padding=True, truncation=True, return_tensors='pt')
通常对于检索任务给query加上instruction会更好，当然您可以根据您的使用情况自行判断加不加instruction。

**Comment by cristianohello** [comment]
@staoxiao 

如果我想检索出queries = ['卡塔尔世界杯冠军是谁？', '中国首都是哪']
这个列表中最详细的sentence文本，该怎么修改代码？

输出结果是：
卡塔尔世界杯冠军是谁？：卡塔尔世界杯冠军是？
中国首都是哪：北京是中国首都



from transformers import AutoTokenizer, AutoModel
import torch
# Sentences we want sentence embeddings for

queries = ['卡塔尔世界杯冠军是谁？', '中国首都是哪']
sentences = ["卡塔尔世界杯冠军是？", "北京是中国首都"]
# Load model from HuggingFace Hub
tokenizer = AutoTokenizer.from_pretrained('/root/autodl-tmp/zys/BAAIbge-large-zh')
model = AutoModel.from_pretrained('/root/autodl-tmp/zys/BAAIbge-large-zh')
# Tokenize sentences
#encoded_input = tokenizer(sentences, padding=True, truncation=True, return_tensors='pt')
# for retrieval task, add an instruction to query
instruction = "为这个句子生成表示以用于检索相关文章："

encoded_input = tokenizer([instruction + q for q in queries], padding=True, truncation=True, return_tensors='pt')
# Compute token embeddings
with torch.no_grad():
    model_output = model(**encoded_input)
    # Perform pooling. In this case, cls pooling.
    sentence_embeddings = model_output[0][:, 0]
# normalize embeddings
sentence_embeddings = torch.nn.functional.normalize(sentence_embeddings, p=2, dim=1)
print("Sentence embeddings:", sentence_embeddings)

**Comment by cristianohello** [comment]
@staoxiao 

比如：我想取出这个passages 元素：手机开不了机怎么办啊？？该怎么修改？

`from sentence_transformers import SentenceTransformer
import numpy as np


queries = ["手机开不了机怎么办？"]
passages = ["手机开不了机去哪修？", "手机开不了机怎么办啊？？？"]
instruction = "为这个句子生成表示以用于检索相关文章："
model = SentenceTransformer('/root/autodl-tmp/zys/BAAIbge-large-zh')
q_embeddings = model.encode([instruction+q for q in queries], normalize_embeddings=True)
p_embeddings = model.encode(passages, normalize_embeddings=True)
scores = q_embeddings @ p_embeddings.T
print(scores)

max_index = np.argmax(scores, axis=1)


print("每行的最大值对应的索引：", max_index)

'''
[[0.8127507  0.83004653]]
每行的最大值对应的索引： [1]
'''`

f

**Comment by staoxiao** [comment]
> @staoxiao
> 
> 比如：我想取出这个passages 元素：手机开不了机怎么办啊？？该怎么修改？
> 
> `from sentence_transformers import SentenceTransformer import numpy as np
> 
> queries = ["手机开不了机怎么办？"] passages = ["手机开不了机去哪修？", "手机开不了机怎么办啊？？？"] instruction = "为这个句子生成表示以用于检索相关文章：" model = SentenceTransformer('/root/autodl-tmp/zys/BAAIbge-large-zh') q_embeddings = model.encode([instruction+q for q in queries], normalize_embeddings=True) p_embeddings = model.encode(passages, normalize_embeddings=True) scores = q_embeddings @ p_embeddings.T print(scores)
> 
> max_index = np.argmax(scores, axis=1)
> 
> print("每行的最大值对应的索引：", max_index)
> 
> ''' [[0.8127507 0.83004653]] 每行的最大值对应的索引： [1] '''`
> 
> f

您好，我觉得这个使用没有问题。
对于是否一定要加instruction，大部分检索任务可以默认加上，但最好是根据您的具体使用情况进行判断。
